esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -6590,5 +6591,104 @@ bool CWsDfuEx::onDFUFilePublish(IEspContext &context, IEspDFUFilePublishRequest
     return true;
 }
 
+void CWsDfuEx::setPublishFileSize(const char *lfn, IFileDescriptor *fileDesc)
+{
+    class CFilePartSizeThread : public Thread
+    {
+        StringAttr lfn;
+        IPartDescriptor &part;
+        bool compressed = false;
+        offset_t size = (offset_t)-1;
+        offset_t compressedSize = (offset_t)-1;
+
+    public:
+        CFilePartSizeThread(const char *_lfn, IPartDescriptor &_part, bool _compressed)
+            : lfn(_lfn), part(_part), compressed(_compressed) {};
+
+        virtual int run() override
+        {
+            for (unsigned c = 0; c < part.numCopies(); c++)
+            {
+                RemoteFilename rfn;
+                part.getFilename(c, rfn);
+                try
+                {
+                    Owned<IFile> file = createIFile(rfn);
+                    if (compressed)
+                    {
+                        compressedSize = file->size();
+                        Owned<IFileIO> io = createCompressedFileReader(file);
+                        if (io)
+                            size = io->size();
+                    }
+                    else
+                    {
+                        size = file->size();
+                    }
+                }
+                catch (IException *e)
+                {
+                    VStringBuffer tmp("%s: ", lfn.get());
+                    rfn.getPath(tmp);
+                    EXCLOG(e, tmp.str());
+                    e->Release();
+
+                    size = (offset_t)-1;
+                    compressedSize = (offset_t)-1;
+                }
+                if ((size != (offset_t)-1) && (!compressed || (compressedSize != (offset_t)-1)))
+                    break;
+            }
+            return 0;
+        };
+
+        offset_t getSize() { return size; }
+        offset_t getCompressedSize() { return compressedSize; }
+    };
+
+    //Get the sizes and the compressed sizes for each file parts in parallel
+    CIArrayOf<CFilePartSizeThread> threads;
+    bool compressed = fileDesc->isCompressed();
+    Owned<IPartDescriptorIterator> pi = fileDesc->getIterator();
+    ForEach(*pi)
+    {
+        CFilePartSizeThread* thread = new CFilePartSizeThread(lfn, pi->query(), compressed);
```

Comment:
this will be running an unlimited number of threads, in this context that's ok generally,
but this is a good case for using CAsyncFor (or CAsyncForFunc), e.g. with max 100
and it would significantly reduce code complexity too.

Reply:
I will use CAsyncFor.